### ISSUE TITLE
-sigopt option doesn't work with "nonce-type" in openssl dgst

### ISSUE THREAD

**Issue by tomato42** [openssl/openssl#25012]
Using the `nonce-type:1` as an option to the `openssl dgst` command doesn't work.

Tried 7fc3a59f3f62dc06 (current master) and 3.2.2.

```
openssl genpkey -out key.pem -algorithm EC -pkeyopt ec_paramgen_curve:P-256
echo message to sign > message.txt
openssl dgst -sha256 -sign key.pem -out message.sig1 -sigopt nonce-type:1 message.txt
```
fails with
```
Signature parameter error "nonce-type:1"
40E74E779F7F0000:error:03000093:digital envelope routines:default_fixup_args:command not supported:crypto/evp/ctrl_params_translate.c:578:[action:2, state:4] name=nonce-type, value=1
```

**Comment by tomato42** [comment]
The plot thickens: so it is possible to use deterministic ECDSA using `openssl pkeyutl`, _but_ specifying `nonce-type` alone is not sufficient:
```
openssl dgst -binary -out message.dgst message.txt
openssl pkeyutl -sign -inkey key.pem -pkeyopt nonce-type:1 -out message.sig1 -in message.dgst
```
results in:
```
Public Key operation error
40277166557F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:355:Global default library context, Algorithm ( : 0), Properties (<null>)
40277166557F0000:error:0800009E:elliptic curve routines:ecdsa_sign_setup:random number generation failed:crypto/ec/ecdsa_ossl.c:212:
```

It's only the combination of `-pkeyopt digest:sha256 -pkeyopt nonce-type:1` that works:
```
openssl pkeyutl -sign -inkey key.pem -pkeyopt digest:sha256 -pkeyopt nonce-type:1 -out message.sig1 -in message.dgst
```

**But** the same trick doesn't work for `openssl dgst`:
```
openssl dgst -sha256 -sigopt digest:sha256 -sigopt nonce-type:1 -sign key.pem -out message.sig message.txt
```
aborts with
```
Signature parameter error "digest:sha256"
40177E15BC7F0000:error:03000093:digital envelope routines:default_fixup_args:command not supported:crypto/evp/ctrl_params_translate.c:578:[action:2, state:4] name=digest, value=sha256
```

**Comment by mattcaswell** [comment]
> The plot thickens: so it is possible to use deterministic ECDSA using openssl pkeyutl, but specifying nonce-type alone is not sufficient:

This is in fact documented...but not easy to find. Pages mentioning the `nonce-type` option refer to this page:

https://docs.openssl.org/master/man7/provider-signature/#signature-parameters

Which says this about the `nonce-type` option:

`Set this to 1 to use deterministic digital signature generation with ECDSA or DSA, as defined in RFC 6979 (see Section 3.2 "Generation of k"). In this case, the "digest" parameter must be explicitly set (otherwise, deterministic nonce generation will fail).`

I find the error slightly curious though:

`40277166557F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:355:Global default library context, Algorithm ( : 0), Properties (<null>)`

In particular `Algorithm ( : 0)` implies that a digest of `""` (i.e. the empty string) was passed, instead of a NULL pointer. I might wonder that if we had passed a NULL pointer whether the code might have detected the lack of digest earlier and given a better and more informative error message??

**Comment by mattcaswell** [comment]
Confirmed as a bug. Fix in #25057
